src/chocolatey.console/Program.cs

```diff
@@ -238,20 +238,45 @@ private static void add_assembly_resolver()
 #else
                 var chocolateyPublicKey = ApplicationParameters.UnofficialChocolateyPublicKey;
 #endif
+                if (!requestedAssembly.get_public_key_token().is_equal_to(chocolateyPublicKey))
+                {
+                    // This resolver is only loading official extensions for Chocolatey and ILMerged assemblies.
+                    // Everything else is not handled by this resolver.
+                    return null;
+                }
+
+                // Check if the requested assembly is already loaded
+                var resolvedAssembly = AssemblyResolution.resolve_existing_assembly(requestedAssembly.Name, chocolateyPublicKey);
 
-                if (requestedAssembly.get_public_key_token().is_equal_to(chocolateyPublicKey))
+                if (resolvedAssembly != null)
                 {
-                    // Check if it is already loaded
-                    var resolvedAssembly = AssemblyResolution.resolve_existing_assembly(requestedAssembly.Name, chocolateyPublicKey);
+                    return resolvedAssembly.UnderlyingType;
+                }
 
-                    if (resolvedAssembly != null)
+                if (requestedAssembly.Name.is_equal_to(ApplicationParameters.LicensedChocolateyAssemblySimpleName))
+                {
+                    try
                     {
-                        return resolvedAssembly.UnderlyingType;
+                        "chocolatey".Log().Debug(() => "Resolving reference to chocolatey.licensed...");
+
+                        return AssemblyResolution.load_assembly(
+                            ApplicationParameters.LicensedChocolateyAssemblySimpleName,
+                            ApplicationParameters.LicensedAssemblyLocation,
+                            requestedAssembly.get_public_key_token()).UnderlyingType;
                     }
+                    catch (Exception ex)
+                    {
+                        "chocolatey".Log().Warn("Unable to load chocolatey.licensed assembly. {0}".format_with(ex.Message));
+                        return null;
+                    }
+                }
 
-                    if (Directory.Exists(ApplicationParameters.ExtensionsLocation))
+                if (Directory.Exists(ApplicationParameters.ExtensionsLocation))
+                {
+                    var extensions = Directory.EnumerateFiles(ApplicationParameters.ExtensionsLocation, requestedAssembly.Name + ".dll", SearchOption.AllDirectories).ToList();

```

Comment:
Don't use `.ToList()` here. We rely on lazy enumeration in case there are/will be multiple dll files in the future.

Doing `ToList` adds all assemblies into memory (the string at least), while not doing it will load them one by one.

Reply:
The ToList() is needed so we can check the Count before enumerating. This is done so that if we know we found something here that looks like it should be an extension but it can't be loaded, we ensure that we don't continue to the fallback which is to return the `chocolatey` assembly itself. 

That behaviour _was_ the bug this is fixing. This is necessary to fix the bug.

Reply:
I am aware, which is why had suggested using a boolean value to determine whether we had gone into the foreach loop, instead of calling `ToList` and checking its Count value.

Using a boolean value to determine the need has a lower overhead than having to call `ToList` every time.

Bottom line, if going down this route, change the `EnumerateFiles` to `GetFiles` instead, which returns a pre-allocated array that can be used (I would suggest going with a for loop instead of a foreach loop in that case though for less overhead).

Reply:
Yeah, using a bool is a good idea, thanks. I'll go with that I think.